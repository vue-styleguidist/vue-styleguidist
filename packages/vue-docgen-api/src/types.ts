import type { Options as PugOptions } from 'pug'
import type { Descriptor } from './Documentation'
import type { Handler as ScriptHandler } from './parse-script'
import type { Handler as TemplateHandler } from './parse-template'

export interface ParseOptions extends DocGenOptions, Descriptor {
	validExtends: (fullFilePath: string) => boolean
	filePath: string
	/**
	 * In what language is the component written
	 * @default undefined - let the system decide
	 */
	lang?: 'ts' | 'js'
}

export interface DocGenOptions {
	/**
	 * Which exported variables should be looked at
	 * @default undefined - means treat all exports
	 */
	nameFilter?: string[]
	/**
	 * What alias should be replaced in requires and imports
	 */
	alias?: { [alias: string]: string | string[] }
	/**
	 * What directories should be searched when resolving modules
	 */
	modules?: string[]
	/**
	 * Handlers that will be added at the end of the script analysis
	 */
	addScriptHandlers?: ScriptHandler[]
	/**
	 * Handlers that will be added at the end of the template analysis
	 */
	addTemplateHandlers?: TemplateHandler[]
	/**
	 * Handlers that will replace the extend and mixins analyzer
	 * They will be run before the main component analysis to avoid bleeding onto the main
	 */
	scriptPreHandlers?: ScriptHandler[]
	/**
	 * Handlers that will replace the main script analysis
	 */
	scriptHandlers?: ScriptHandler[]
	/**
	 * Handlers that will replace the template analysis
	 */
	templateHandlers?: TemplateHandler[]
	/**
	 * Does parsed components use jsx?
	 * @default true - if you do not disable it, babel will fail with `(<any>window).$`
	 */
	jsx?: boolean
	/**
	 * Should extended components be parsed?
	 * @default `fullFilePath=>!/[\\/]node_modules[\\/]/.test(fullFilePath)`
	 */
	validExtends?: (fullFilePath: string) => boolean
	/**
	 * all pug options passed to the pug compiler if you use it
	 */
	pugOptions?: PugOptions
}