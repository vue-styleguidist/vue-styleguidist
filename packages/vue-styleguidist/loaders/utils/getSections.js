// This two functions should be in the same file because of cyclic imports

const fs = require('fs')
const path = require('path')
const _ = require('lodash')
const requireIt = require('react-styleguidist/lib/loaders/utils/requireIt')
const getComponentFiles = require('react-styleguidist/lib/loaders/utils/getComponentFiles')
const slugger = require('react-styleguidist/lib/loaders/utils/slugger')
const getComponents = require('./getComponents')

const examplesLoader = path.resolve(__dirname, '../examples-loader.js')

/**
 * Return object for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @param {number} parentDepth
 * @returns {Array}
 */
function getSections(sections, config, parentDepth) {
	return sections.map(section => processSection(section, config, parentDepth))
}

const getSectionComponents = (section, config) => {
	let ignore = config.ignore ? _.castArray(config.ignore) : []
	if (section.ignore) {
		ignore = ignore.concat(_.castArray(section.ignore))
	}

	return getComponents(getComponentFiles(section.components, config.configDir, ignore), config)
}

/**
 * Return an object for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @param {number} parentDepth
 * @returns {object}
 */
function processSection(section, config, parentDepth) {
	const contentRelativePath = section.content

	// Try to load section content file
	let content
	if (contentRelativePath) {
		const contentAbsolutePath = path.resolve(config.configDir, contentRelativePath)
		if (!fs.existsSync(contentAbsolutePath)) {
			throw new Error(`Styleguidist: Section content file not found: ${contentAbsolutePath}`)
		}
		content = requireIt(`!!${examplesLoader}?customLangs=vue|js|jsx!${contentAbsolutePath}`)
	}

	let sectionDepth

	if (parentDepth === undefined) {
		sectionDepth = section.sectionDepth !== undefined ? section.sectionDepth : 0
	} else {
		sectionDepth = parentDepth === 0 ? 0 : parentDepth - 1
	}

	return {
		name: section.name,
		exampleMode: section.exampleMode || config.exampleMode,
		usageMode: section.usageMode || config.usageMode,
		sectionDepth,
		description: section.description,
		slug: slugger.slug(section.name),
		sections: getSections(section.sections || [], config, sectionDepth),
		filepath: contentRelativePath,
		href: section.href,
		components: getSectionComponents(section, config),
		content,
		external: section.external
	}
}

module.exports = getSections
module.exports.processSection = processSection
