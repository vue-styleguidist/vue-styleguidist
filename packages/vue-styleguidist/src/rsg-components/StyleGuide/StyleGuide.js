import React, { Component } from 'react'
import PropTypes from 'prop-types'
import TableOfContents from 'rsg-components/TableOfContents'
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer'
import { DisplayModes } from 'react-styleguidist/lib/client/consts'
import Sections from 'rsg-components/Sections'
import Welcome from 'rsg-components/Welcome'
import Error from 'rsg-components/Error'
import NotFound from 'rsg-components/NotFound'
import { HOMEPAGE } from '../../../scripts/consts'

/**
 * This function will return true, if the sidebar should be visible and false otherwise.
 *
 * These sorted conditions (highest precedence first) define the visibility
 * state of the sidebar.
 *
 * - Sidebar is hidden for isolated example views
 * - Sidebar is always visible when pagePerSection
 * - Sidebar is hidden when showSidebar is set to false
 * - Sidebar is visible when showSidebar is set to true for non-isolated views
 *
 * @param {boolean} displayMode
 * @param {boolean} showSidebar
 * @param {boolean} pagePerSection
 * @returns {boolean}
 */
function hasSidebar(displayMode, showSidebar) {
	return displayMode === DisplayModes.notFound || (showSidebar && displayMode === DisplayModes.all)
}

export default class StyleGuide extends Component {
	static propTypes = {
		codeRevision: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		renderRootJsx: PropTypes.object,
		patterns: PropTypes.array,
		displayMode: PropTypes.string,
		allSections: PropTypes.array.isRequired,
		pagePerSection: PropTypes.bool
	}

	static childContextTypes = {
		codeRevision: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		displayMode: PropTypes.string,
		renderRootJsx: PropTypes.object
	}

	static defaultProps = {
		displayMode: DisplayModes.all
	}

	state = {
		error: false,
		info: null
	}

	getChildContext() {
		return {
			codeRevision: this.props.codeRevision,
			config: this.props.config,
			slots: this.props.slots,
			displayMode: this.props.displayMode,
			renderRootJsx: this.props.renderRootJsx
		}
	}

	componentDidCatch(error, info) {
		this.setState({
			error,
			info
		})
	}

	render() {
		const {
			config,
			sections,
			welcomeScreen,
			patterns,
			displayMode,
			allSections,
			pagePerSection
		} = this.props

		if (this.state.error) {
			return <Error error={this.state.error} info={this.state.info} />
		}

		if (welcomeScreen) {
			return <Welcome patterns={patterns} />
		}

		return (
			<StyleGuideRenderer
				title={config.title}
				version={config.version}
				homepageUrl={HOMEPAGE}
				toc={<TableOfContents sections={allSections} useRouterLinks={pagePerSection} />}
				hasSidebar={hasSidebar(displayMode, config.showSidebar)}
			>
				{sections.length ? <Sections sections={sections} depth={1} /> : <NotFound />}
			</StyleGuideRenderer>
		)
	}
}
