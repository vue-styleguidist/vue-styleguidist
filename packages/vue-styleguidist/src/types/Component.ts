import 'react-styleguidist/lib/typings/dependencies/react-docgen'

import { VueConstructor } from 'vue'
import {
	PropDescriptor,
	MethodDescriptor,
	SlotDescriptor,
	EventDescriptor,
	BlockTag
} from 'vue-docgen-api'
import * as Rsg from 'react-styleguidist'
import { CodeExample } from './Example'

interface BaseComponentProps {
	displayName: string
	description?: string
	props?: PropDescriptor[]
	methods?: MethodDescriptor[]
	slots?: { [name: string]: SlotDescriptor }
	events?: { [name: string]: EventDescriptor }
	tags?: {
		[key: string]: BlockTag[]
	}
	docsBlocks?: string[]
	visibleName?: string
	examplesFile?: string
}

export interface LoaderComponentProps extends BaseComponentProps {
	examples?: Rsg.RequireItResult | null
	example?: Rsg.RequireItResult | Rsg.RequireItResult[] | null
	subComponents?: Rsg.LoaderComponent[]
}

export interface LoaderComponent extends Omit<Rsg.LoaderComponent, 'props'> {
	props: LoaderComponentProps
}

export interface ComponentProps extends BaseComponentProps {
	examples?: (CodeExample | Rsg.MarkdownExample)[]
	example?: (CodeExample | Rsg.MarkdownExample)[] | (CodeExample | Rsg.MarkdownExample)[][]
	subComponents?: Component[]
}

export interface Component extends Omit<Rsg.Component, 'props' | 'module'> {
	props: ComponentProps
	module: { default: VueConstructor } | VueConstructor
}
