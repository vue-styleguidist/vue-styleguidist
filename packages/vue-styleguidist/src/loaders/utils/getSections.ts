// These two functions should be in the same file because of cyclic imports

import * as fs from 'fs'
import * as path from 'path'
import { castArray, flatten } from 'lodash'
import * as Rsg from 'react-styleguidist'
import { parseMulti, ParamTag, ScriptHandlers } from 'vue-docgen-api'
import requireIt from 'react-styleguidist/lib/loaders/utils/requireIt'
import getComponentFiles from 'react-styleguidist/lib/loaders/utils/getComponentFiles'
import slugger from 'react-styleguidist/lib/loaders/utils/slugger'
import { SanitizedStyleguidistConfig } from '../../types/StyleGuide'
import getComponents from './getComponents'

const examplesLoader = path.resolve(__dirname, '../examples-loader.js')

/**
 * Return object for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @param {number} parentDepth
 * @returns {Array}
 */
export default async function getSections(
	sections: Rsg.ConfigSection[],
	config: SanitizedStyleguidistConfig,
	componentFiles: string[],
	requiredComponents?: string[],
	parentDepth?: number
): Promise<Rsg.LoaderSection[]> {
	return Promise.all(
		sections.map(
			async section =>
				await processSection(
					section,
					config,
					componentFiles,
					requiredComponents ||
						(await getRequiredComponents(componentFiles, config.jsxInComponents)),
					parentDepth
				)
		)
	)
}

/**
 * returns all the `@required` file path in the analyzed components
 * this way we can ignore them when analyzing the components in the menu
 * and only add them as a subsection for a parent component
 * @param componentFiles all the component file paths to be analyzed
 */
export async function getRequiredComponents(
	componentFiles: string[],
	jsx: boolean
): Promise<string[]> {
	return flatten(
		await Promise.all(
			componentFiles.map(async componentPath => {
				const compDirName = path.dirname(componentPath)
				const docs = await parseMulti(componentPath, {
					scriptPreHandlers: [],
					scriptHandlers: [ScriptHandlers.componentHandler],
					jsx
				})
				return docs.reduce((acc: string[], doc) => {
					if (doc.tags && doc.tags.requires) {
						acc = acc.concat(
							doc.tags.requires.map((t: ParamTag) =>
								path.resolve(compDirName, t.description as string)
							)
						)
					}
					return acc
				}, [])
			})
		)
	)
}

/**
 * Return an object for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @param {number} parentDepth
 * @returns {object}
 */
export async function processSection(
	section: Rsg.ConfigSection,
	config: SanitizedStyleguidistConfig,
	componentFiles: string[],
	requiredComponents: string[],
	parentDepth?: number
): Promise<Rsg.LoaderSection> {
	const contentRelativePath = section.content

	// Try to load section content file
	let content
	if (contentRelativePath) {
		const contentAbsolutePath = path.resolve(config.configDir, contentRelativePath)
		if (!fs.existsSync(contentAbsolutePath)) {
			throw new Error(`Styleguidist: Section content file not found: ${contentAbsolutePath}`)
		}
		content = requireIt(`!!${examplesLoader}?customLangs=vue|js|jsx!${contentAbsolutePath}`)
	}

	let sectionDepth

	if (parentDepth === undefined) {
		sectionDepth = section.sectionDepth !== undefined ? section.sectionDepth : 0
	} else {
		sectionDepth = parentDepth === 0 ? 0 : parentDepth - 1
	}

	return {
		name: section.name || '',
		exampleMode: section.exampleMode || config.exampleMode,
		usageMode: section.usageMode || config.usageMode,
		sectionDepth,
		description: section.description,
		slug: slugger.slug(section.name || ''),
		sections: await getSections(
			section.sections || [],
			config,
			componentFiles,
			requiredComponents,
			sectionDepth
		),
		href: section.href,
		components: getSectionComponents(section, config, requiredComponents),
		content,
		external: section.external
	}
}

const getSectionComponents = (
	section: Rsg.ConfigSection,
	config: SanitizedStyleguidistConfig,
	requiredComponents: string[]
): Rsg.LoaderComponent[] => {
	let ignore = config.ignore ? castArray(config.ignore) : []
	if (section.ignore) {
		ignore = ignore.concat(castArray(section.ignore))
	}
	if (requiredComponents.length) {
		ignore = ignore.concat(requiredComponents)
	}

	return getComponents(getComponentFiles(section.components, config.configDir, ignore), config)
}
