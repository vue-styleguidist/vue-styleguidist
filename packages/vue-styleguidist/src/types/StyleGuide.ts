/**
 * /!\ WARNING /!\
 * Do not edit manually.
 * This file is the compilation of
 * Template: packages/vue-styleguidist/templates/StyleGuide.ts.ejs
 * Config Data: packages/vue-styleguidist/src/scripts/schemas/config.ts
 */

import React from 'react'
import WebpackDevServer from 'webpack-dev-server'
import { ComponentDoc, PropDescriptor } from 'vue-docgen-api'
import { TransformOptions } from 'buble'
import { Styles } from 'jss'
import { Configuration, loader } from 'webpack'
import * as Rsg from 'react-styleguidist'
import { RecursivePartial } from 'react-styleguidist/lib/typings/RecursivePartial'
import { ProcessedSection } from './Section'
import { LoaderComponentProps } from './Component'

export interface StyleguidistContext extends loader.LoaderContext {
	_styleguidist: SanitizedStyleguidistConfig
}

export interface BaseStyleguidistConfig
	extends Omit<
		Rsg.SanitizedStyleguidistConfig,
		'sections' | 'propsParser' | 'sortProps' | 'updateDocs'
	> {
	/**
	 * Your application static assets folder, will be accessible as / in the style guide dev server.
	 */
	assetsDir: string
	/**
	 * Should the styleguide try code splitting for better performance? NOte that you will need the proper transform in your babel config
	 * @default true
	 */
	codeSplit: boolean
	compilerConfig: TransformOptions
	/**
	 * Where to find the components. Takes in a String or an Array of glob paths. Comma separated.
	 */
	components: (() => string[]) | string | string[]
	configDir: string
	context: Record<string, any>
	contextDependencies: string[]
	configureServer: (server: WebpackDevServer, env: string) => string
	/**
	 * Add a button on the top right of the code sections to copy to clipboard the current contents of the editor
	 * @default false
	 */
	copyCodeButton: boolean
	/**
	 * Allows you to modify webpack config without any restrictions
	 */
	dangerouslyUpdateWebpackConfig: (server: Configuration, env: string) => Configuration
	/**
	 * Display each component with a default example, regardless of if there's a README or <docs/> block written.
	 * @default false
	 */
	defaultExample: string
	/**
	 * In the generated docs, this adda a column to the props table giving in which file it is defined. Useful when extending components or mixing mixins
	 * @default false
	 */
	displayOrigins: boolean
	editorConfig: {
		theme: string
	}
	/**
	 * Defines the initial state of the props and methods tab
	 * @default "collapse"
	 */
	exampleMode: Rsg.ExpandMode
	getComponentPathLine: (componentPath: string) => string
	getExampleFilename: (componentPath: string) => string
	/**
	 * @deprecated Use the theme property in the editorConfig option instead
	 * @default "base16-light"
	 */
	highlightTheme: string
	/**
	 * What components to ignore. Can be an Array or String. Comma separated.
	 */
	ignore: string[]
	/**
	 * By default, the PrismJs editor is themed in the theme files. If you want to use a theme defined in CSS, set this to false and require the CSS file in the `require` config.
	 * @default true
	 */
	jssThemedEditor: boolean
	/**
	 * Do documented components contain JSX syntax? Set this to `false` to restore compatibility with this TypeScript cast syntax: `<any>variable` instead of `variable as any`.
	 * @default true
	 */
	jsxInComponents: boolean
	/**
	 * Allow exmaples to contain JSX syntax. Use proper JSX Vue component format in examples.
	 * @default false
	 */
	jsxInExamples: boolean
	/**
	 * Register components on their examples only instead of globally Vue.components()
	 * @default false
	 */
	locallyRegisterComponents: boolean
	/**
	 * @deprecated Use pagePerSection option instead
	 * @default false
	 */
	navigation: boolean
	/**
	 * @deprecated Use renderRootJsx option instead
	 */
	mixins: string[]
	logger: {
		info(message: string): void
		warn(message: string): void
		debug(message: string): void
	}
	/**
	 * If this option is set to false, the styelguidist will not minimize the js at build.
	 * @default true
	 */
	minimize: boolean
	/**
	 * The ID of a DOM element where Styleguidist mounts.
	 * @default "rsg-root"
	 */
	mountPointId: string
	/**
	 * Render one section or component per page. If true, each section will be a single page.
	 * @default false
	 */
	pagePerSection: boolean
	/**
	 * @default 500
	 */
	previewDelay: number
	printBuildInstructions: any
	printServerInstructions: any
	/**
	 * Display a progress bar while building
	 * @default true
	 */
	progressBar: boolean
	propsParser: (file: string) => Promise<ComponentDoc>
	require: string[]
	renderRootJsx: string
	/**
	 * Shows 'Fork Me' ribbon in the top-right corner. If ribbon key is present, then it's required to add url property; text property is optional. If you want to change styling of the ribbon, please, refer to the theme section in the documentation.
	 */
	ribbon: {
		url: string
		text: string
	}
	/**
	 * Dev server host name
	 * @default "0.0.0.0"
	 */
	serverHost: string
	/**
	 * Dev server port
	 * @default 6060
	 */
	serverPort: number
	/**
	 * @deprecated Use exampleMode option instead
	 * @default false
	 */
	showCode: boolean
	/**
	 * @deprecated Use usageMode option instead
	 * @default false
	 */
	showUsage: boolean
	/**
	 * Toggle sidebar visibility. Sidebar will be hidden when opening components or examples in isolation mode even if this value is set to true. When set to false, sidebar will always be hidden.
	 * @default true
	 */
	showSidebar: boolean
	/**
	 * Avoid loading CodeMirror and reduce bundle size significantly, use prism.js for code highlighting. Warning: editor options will not be mapped over.
	 * @default true
	 */
	simpleEditor: boolean
	/**
	 * Ignore components that don’t have an example file (as determined by getExampleFilename). These components won’t be accessible from other examples unless you manually require them.
	 * @default false
	 */
	skipComponentsWithoutExample: boolean
	sortProps: (props: PropDescriptor[]) => PropDescriptor[]
	styleguideComponents: { [name: string]: string }
	/**
	 * Folder for static HTML style guide generated with `styleguidist build` command.
	 * @default "styleguide"
	 */
	styleguideDir: string
	/**
	 * configures the prefix of the server and built urls.
	 * @default ""
	 */
	styleguidePublicPath: string
	styles: Styles | string | ((theme: any) => Styles)
	template: any
	theme: { [name: string]: any } | string
	/**
	 * Style guide title
	 */
	title: string
	updateDocs: (doc: LoaderComponentProps, file: string) => LoaderComponentProps
	updateExample: (
		props: Pick<Rsg.CodeExample, 'content' | 'lang' | 'settings'>,
		ressourcePath: string
	) => Rsg.CodeExample
	updateWebpackConfig: any
	/**
	 * Defines the initial state of the props and methods tab
	 * @default "collapse"
	 */
	usageMode: Rsg.ExpandMode
	/**
	 * If set to collapse, the sidebar sections are collapsed by default. Handy when dealing with big Components bases
	 * @default "expand"
	 */
	tocMode: Rsg.ExpandMode
	/**
	 * Should the passed filepath be parsed by docgen if mentionned extends
	 */
	validExtends: (filePath: string) => boolean
	/**
	 * Print debug information. Same as --verbose command line switch.
	 * @default false
	 */
	verbose: boolean
	/**
	 * The version # of the Styleguide
	 */
	version: string
	/**
	 * @deprecated Use renderRootJsx option instead
	 */
	vuex: any
	webpackConfig: Configuration
}

export interface SanitizedStyleguidistConfig extends BaseStyleguidistConfig {
	sections: Rsg.ConfigSection[]
}

/**
 * definition of the config object where everything is optional
 * note that teh default example can be both a string and a boolean but ends
 * up only being a string after sanitizing
 */
export interface StyleguidistConfig
	extends RecursivePartial<Omit<SanitizedStyleguidistConfig, 'defaultExample'>> {
	defaultExample?: string | boolean
}

export interface StyleGuideObject {
	sections: ProcessedSection[]
	config: StyleguidistConfig
	renderRootJsx: React.ReactNode
	welcomeScreen: any
	patterns: string[]
}
