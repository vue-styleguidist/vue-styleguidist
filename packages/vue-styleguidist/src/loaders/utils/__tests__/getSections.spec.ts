import path from 'path'
import * as Rsg from 'react-styleguidist'
import getSections, { processSection, getRequiredComponents } from '../getSections'
import { SanitizedStyleguidistConfig } from '../../../types/StyleGuide'

const configDir = path.resolve(__dirname, '../../../../../../test')
const config = {
	configDir,
	exampleMode: 'collapse',
	usageMode: 'collapse',
	getExampleFilename: (a: string) => a,
	getComponentPathLine: (a: string) => a
} as SanitizedStyleguidistConfig

const sections: Rsg.ConfigSection[] = [
	{
		name: 'Readme',
		content: 'components/ReadMe.md'
	},
	{
		name: 'Components',
		components: 'components/**/[A-Z]*.vue'
	},
	{
		name: 'Ignore',
		components: 'components/**/*.vue',
		ignore: '**/components/Annotation/*'
	}
]

const sectionsWithDepth = [
	{
		name: 'Documentation',
		sections: [
			{
				name: 'Files',
				sections: [
					{
						name: 'First File'
					}
				]
			}
		],
		sectionDepth: 2
	},
	{
		name: 'Components',
		sections: [
			{
				name: 'Buttons'
			}
		],
		sectionDepth: 0
	}
]
const sectionsWithBadDepth = [
	{
		name: 'Documentation',
		sections: [
			{
				name: 'Files',
				sections: [
					{
						name: 'First File'
					}
				],
				sectionDepth: 2
			}
		]
	}
]

function filterSectionDepth(section: Rsg.LoaderSection): Rsg.ConfigSection {
	if (section.sections && section.sections.length) {
		return {
			sectionDepth: section.sectionDepth,
			sections: section.sections.map(filterSectionDepth)
		}
	}
	return {
		sectionDepth: section.sectionDepth
	}
}

describe('processSection', () => {
	it('should return an object for section with content', async () => {
		const result = await processSection(sections[0], { config, componentFiles: [] })

		expect(result).toMatchSnapshot()
	})

	it('should return an object for section with components', async () => {
		const result = await processSection(sections[1], { config, componentFiles: [] })

		expect(result).toMatchSnapshot()
	})

	it('should return an object for section without ignored components', async () => {
		const result = await processSection(sections[2], { config, componentFiles: [] })

		expect(result).toMatchSnapshot()
	})
})

describe('getSections', () => {
	it('should return an array', async () => {
		const result = await getSections(sections, { config, componentFiles: [] })

		expect(result).toMatchSnapshot()
	})

	it('should return an array of sectionsWithDepth with sectionDepth decreasing', async () => {
		const result = await getSections(sectionsWithDepth, { config, componentFiles: [] })

		expect(result.map(filterSectionDepth)).toEqual([
			{
				sectionDepth: 2,
				sections: [
					{
						sectionDepth: 1,
						sections: [
							{
								sectionDepth: 0
							}
						]
					}
				]
			},
			{
				sectionDepth: 0,
				sections: [
					{
						sectionDepth: 0
					}
				]
			}
		])
	})

	it('should return an array of sectionsWithBadDepth taking the sectionDepth of the first depth of the sections', async () => {
		const result = await getSections(sectionsWithBadDepth, { config, componentFiles: [] })

		expect(result.map(filterSectionDepth)).toEqual([
			{
				sectionDepth: 0,
				sections: [
					{
						sectionDepth: 0,
						sections: [
							{
								sectionDepth: 0
							}
						]
					}
				]
			}
		])
	})
})

jest.mock('vue-docgen-api', () => ({
	parseMulti: () =>
		Promise.resolve([
			{
				tags: {
					requires: [
						{ description: 'path/to/require1' }, //
						{ description: 'path/to/require2' } //
					]
				}
			}
		]),
	ScriptHandlers: {}
}))

describe('getRequiredComponents', () => {
	it('should return an array of all requires tags contents', async () => {
		const requiredFiles = await getRequiredComponents(['source/of/file'], false)
		expect(requiredFiles).toMatchInlineSnapshot(`
		Object {
		  "source/of/file": Array [
		    "~/source/of/path/to/require1",
		    "~/source/of/path/to/require2",
		  ],
		}
	`)
	})
})
