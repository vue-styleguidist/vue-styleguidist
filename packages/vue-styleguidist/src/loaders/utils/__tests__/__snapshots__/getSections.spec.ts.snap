// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSections should return an array 1`] = `
Array [
  Object {
    "components": Array [],
    "content": Object {
      "require": "!!~/packages/vue-styleguidist/src/loaders/examples-loader.js?customLangs=vue|js|jsx!~/test/components/ReadMe.md",
    },
    "description": undefined,
    "exampleMode": "collapse",
    "external": undefined,
    "href": undefined,
    "name": "Readme",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "readme-1",
    "usageMode": "collapse",
  },
  Object {
    "components": Array [
      Object {
        "filepath": "components/Button.vue",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Button.vue",
        },
        "pathLine": "components/Button.vue",
        "props": Object {
          "require": "!!~/packages/vue-styleguidist/src/loaders/vuedoc-loader.js!~/test/components/Button.vue",
        },
        "slug": "button-2",
      },
    ],
    "content": undefined,
    "description": undefined,
    "exampleMode": "collapse",
    "external": undefined,
    "href": undefined,
    "name": "Components",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "components-1",
    "usageMode": "collapse",
  },
  Object {
    "components": Array [
      Object {
        "filepath": "components/Button.vue",
        "hasExamples": true,
        "metadata": Object {},
        "module": Object {
          "require": "~/test/components/Button.vue",
        },
        "pathLine": "components/Button.vue",
        "props": Object {
          "require": "!!~/packages/vue-styleguidist/src/loaders/vuedoc-loader.js!~/test/components/Button.vue",
        },
        "slug": "button-3",
      },
    ],
    "content": undefined,
    "description": undefined,
    "exampleMode": "collapse",
    "external": undefined,
    "href": undefined,
    "name": "Ignore",
    "sectionDepth": 0,
    "sections": Array [],
    "slug": "ignore-1",
    "usageMode": "collapse",
  },
]
`;

exports[`processSection should return an object for section with components 1`] = `
Object {
  "components": Array [
    Object {
      "filepath": "components/Button.vue",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Button.vue",
      },
      "pathLine": "components/Button.vue",
      "props": Object {
        "require": "!!~/packages/vue-styleguidist/src/loaders/vuedoc-loader.js!~/test/components/Button.vue",
      },
      "slug": "button",
    },
  ],
  "content": undefined,
  "description": undefined,
  "exampleMode": "collapse",
  "external": undefined,
  "href": undefined,
  "name": "Components",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "components",
  "usageMode": "collapse",
}
`;

exports[`processSection should return an object for section with content 1`] = `
Object {
  "components": Array [],
  "content": Object {
    "require": "!!~/packages/vue-styleguidist/src/loaders/examples-loader.js?customLangs=vue|js|jsx!~/test/components/ReadMe.md",
  },
  "description": undefined,
  "exampleMode": "collapse",
  "external": undefined,
  "href": undefined,
  "name": "Readme",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "readme",
  "usageMode": "collapse",
}
`;

exports[`processSection should return an object for section without ignored components 1`] = `
Object {
  "components": Array [
    Object {
      "filepath": "components/Button.vue",
      "hasExamples": true,
      "metadata": Object {},
      "module": Object {
        "require": "~/test/components/Button.vue",
      },
      "pathLine": "components/Button.vue",
      "props": Object {
        "require": "!!~/packages/vue-styleguidist/src/loaders/vuedoc-loader.js!~/test/components/Button.vue",
      },
      "slug": "button-1",
    },
  ],
  "content": undefined,
  "description": undefined,
  "exampleMode": "collapse",
  "external": undefined,
  "href": undefined,
  "name": "Ignore",
  "sectionDepth": 0,
  "sections": Array [],
  "slug": "ignore",
  "usageMode": "collapse",
}
`;
